### PR TITLE
Generate AST from code

### LINKED ISSUE #167: Research: Kotlin can generate AST from any part of code
Let's create mechanism that will be doing the following:
1) You will provide a small code snippet, for example: 
```
a = foo(15)
```

Kotlin compiler will not be able to parse it. 

2) BUT! Koltin is able to build a code from this text block and will generate a sub tree. This tree can be injected in any other Kotlin Tree without any generation of such mechanisms. 

3) We need Util methods that will be doing that from a string code.

### PR DESCRIPTION
**Pull request by kentr0w**
### What's done:
   Build code from text and generate a sub tree.
   closes #167

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringToNode.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import org.jetbrains.kotlin.com.intellij.openapi.Disposable
+import org.jetbrains.kotlin.config.CompilerConfiguration
+import org.jetbrains.kotlin.psi.KtPsiFactory
+
+class StringToNode {
+
+    /**
+     * Set property
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringToNode.kt:15]: Please provide more details/links regarding which error it helps to workaround and so on.

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/utils/StringUtilsTest.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        val node  = StringToNode().createNode("val x: Int = 10")!!
+        Assert.assertEquals(PROPERTY, node.elementType)
+        Assert.assertEquals("val x: Int = 10", node.text)
+        Assert.assertEquals(4, node.findAllNodesWithSpecificType(WHITE_SPACE).size)
+    }
+
+    @Test
+    fun `test property with expression`() {
+        val node  = StringToNode().createNode("val x = foo(15)")!!
+        println(node.prettyPrint())
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/utils/StringUtilsTest.kt:23]: Please remove

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/utils/StringUtilsTest.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        Assert.assertEquals(PROPERTY, node.elementType)
+        Assert.assertEquals("val x = foo(15)", node.text)
+        Assert.assertEquals("foo(15)", node.findChildByType(CALL_EXPRESSION)!!.text)
+        Assert.assertEquals(3, node.findAllNodesWithSpecificType(WHITE_SPACE).size)
+    }
+
+    @Test
+    fun `test inline function`() {
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/utils/StringUtilsTest.kt:31]: This is not an inline function in your example

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringToNode.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    fun createNode(text: String): ASTNode? {
+        val project = KotlinCoreEnvironment.createForProduction(
+                Disposable {},
+                CompilerConfiguration(),
+                EnvironmentConfigFiles.JVM_CONFIG_FILES
+        ).project
+        val ktPsiFactory = KtPsiFactory(project, true)
+        val node = ktPsiFactory.createBlockCodeFragment(text, null).node
+        return node.findChildByType(BLOCK)?.firstChildNode
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringToNode.kt:29]: How can it be nullable? Also, I think if supplied code is not valid (i.e. cannot be parsed) we should at least log error message, or even throw an exception.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringToNode.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import com.pinterest.ktlint.core.ast.ElementType.BLOCK
+import org.jetbrains.kotlin.cli.common.environment.setIdeaIoUseFallback
+import org.jetbrains.kotlin.cli.jvm.compiler.EnvironmentConfigFiles
+import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import org.jetbrains.kotlin.com.intellij.openapi.Disposable
+import org.jetbrains.kotlin.config.CompilerConfiguration
+import org.jetbrains.kotlin.psi.KtPsiFactory
+
+class StringToNode {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringToNode.kt:12]: I suggest to rename this class to something like `ParsingUtils` or `KotlinParser` or something.

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/utils/StringUtilsTest.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package org.cqfn.diktat.ruleset.utils
+
+import com.pinterest.ktlint.core.ast.ElementType.CALL_EXPRESSION
+import com.pinterest.ktlint.core.ast.ElementType.FUN
+import com.pinterest.ktlint.core.ast.ElementType.PROPERTY
+import com.pinterest.ktlint.core.ast.ElementType.WHITE_SPACE
+import org.junit.Assert
+import org.junit.jupiter.api.Test
+
+class StringUtilsTest {
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/utils/StringUtilsTest.kt:10]: Please provide some examples with parsing of multiline code fragments, invalid code fragments.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            throw KotlinParseException("Your text is not valid")
+        val ktPsiFactory = KtPsiFactory(project, true)
+        if (text.trim().isEmpty())
+            return ktPsiFactory.createWhiteSpace(text).node
+        lateinit var node: ASTNode
+        node = if (isPackage)
+            ktPsiFactory.createFile(text).node
+        else
+            ktPsiFactory.createBlockCodeFragment(text, null).node.findChildByType(BLOCK)!!
+        if (node.getChildren(null).size == ONE_CHILD)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt:73]: Why this strange constant? It is allowed to use `1`, it is not considered as a magic number.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+     */
+    init {
+        setIdeaIoUseFallback()
+    }
+
+    companion object{
+        private const val ONE_CHILD = 1
+    }
+
+    fun createNode(text: String, isPackage: Boolean = false): ASTNode {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt:37]: Please add documentation to this class, including some not obvious things we discussed, e.g. `createFile` is called only for `isPackage`.

**Review by petertrr — CHANGES_REQUESTED** [review]
Another important task for this parser is ability to insert generated nodes into AST provided by framework. Actually, this is the main goal we had starting this.

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParserTest.kt [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    @Test
+    fun `test invalidate code`() {
+        assertThrows<KotlinParseException> { KotlinParser().createNode("simple text") }
+        assertThrows<KotlinParseException> { KotlinParser().createNode("") }
+        assertThrows<KotlinParseException> { KotlinParser().createNode("fuc fun() = 1") }
+    }
+
+    @Test
```
Comment (akuleshov7) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParserTest.kt:45]: please add separate tests for package names, separate test for imports

**Review by akuleshov7 — APPROVED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        project.registerService(PomModel::class.java, pomModel)
+        if (text.isEmpty())
+            return null
+        val ktPsiFactory = KtPsiFactory(project, true)
+        if (text.trim().isEmpty())
+            return ktPsiFactory.createWhiteSpace(text).node
+        lateinit var node: ASTNode
+        node = if (isPackage)
+            ktPsiFactory.createFile(text).node
+        else {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt:80]:
```suggestion
        node = if (isPackage) {
            ktPsiFactory.createFile(text).node
        } else {
```
Please add braces, it's really hard to read complex expressions without them.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        ).project // create project
+        project as MockProject
+        project.registerService(PomModel::class.java, pomModel)
+        if (text.isEmpty())
+            return null
+        val ktPsiFactory = KtPsiFactory(project, true)
+        if (text.trim().isEmpty())
+            return ktPsiFactory.createWhiteSpace(text).node
+        lateinit var node: ASTNode
+        node = if (isPackage)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt:78]:
```suggestion
        val node = if (isPackage)
```
Why can't you do like this?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+            ktPsiFactory.createFile(text).node
+        else {
+            if (text.contains(KtTokens.IMPORT_KEYWORD.value)) {
+                val (imports, blockCode) = text.lines().partition { it.contains(KtTokens.IMPORT_KEYWORD.value) }
+                if (blockCode.isNotEmpty())
+                    return null
+                if (imports.size == 1) {
+                    val importText = ImportPath.fromString(text.substringAfter("$IMPORT_KEYWORD "))
+                    ktPsiFactory.createImportDirective(importText).node
+                } else {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt:88]: Will this work for several imports? There should be node with type `IMPORT_LIST` in this case.

**Review by petertrr — COMMENTED** [review]
Could you try how your method works with KDocs? It would be useful for #74, for example. String with KDoc should be parsed as `KDOC` node with all it's complex structure: `KDOC_TEXT`, `LEADING_ASTERISK` etc.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt [hunk 13]
```diff
@@ -104,4 +104,7 @@ class KotlinParser {
         }
         return node
     }
+
+    private fun isContainKDoc(text: String) =
+            text.lines().any { it.trim().indexOf("/**") == 0 }
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt:109]:
```suggestion
            text.lines().any { it.trim().startsWith("/**") }
```

**Review by petertrr — APPROVED** [review]
LGTM, let's merge it and try to use.

**Comment by akuleshov7** [comment]
> Another important task for this parser is ability to insert generated nodes into AST provided by framework. Actually, this is the main goal we had starting this.

@kentr0w did you fix this? create an issue if not

**Pull request merged**